src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs

```diff
@@ -200,7 +201,7 @@ public ReferencedAssemblyIdentity(AssemblyIdentity identity, MetadataReference r
             TCompilation compilation,
             [Out] Dictionary<string, List<ReferencedAssemblyIdentity>> assemblyReferencesBySimpleName,
             out ImmutableArray<MetadataReference> references,
-            out IDictionary<(string, string), MetadataReference> boundReferenceDirectiveMap,
+            [NotNull] out IDictionary<(string, string), MetadataReference>? boundReferenceDirectiveMap,
```

Comment:
what does it do to put 'NotNull' on an 'out' parameter of a reference type?

Reply:
It's like `[NotNullWhen]` except that the "When" portion is always `true`. Basically this method will unconditionally assign a non-null value to the parameter.

Reply:
What I was trying to ask is: why can't this parameter be written as `out IDictionary<(string, string), MetadataReference> boundReferenceDirectiveMap`, instead of the way it is now?

Reply:
Removing the `?` annotation though would mean that I'd have to provide a non-null value at the call site correct? That would break the existing code.

Reply:
I think this is the scenario you are referring to. This doesn't produce a warning because we know the variable we pass using 'out' will only be written to.

```cs
public void M1(string? s1)
{
    M2(out s1); // no warning
    s1.ToString();
}

public void M2(out string s2)
{
    s2 = "a";
}
```

Reply:
Ah okay. Yeah then `[NonNull]` can be removed